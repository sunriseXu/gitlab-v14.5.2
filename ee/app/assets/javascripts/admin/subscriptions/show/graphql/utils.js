import produce from 'immer';
import { subscriptionHistoryQueries, subscriptionQueries } from '../constants';

export const getLicenseFromData = ({ data } = {}) => data?.gitlabSubscriptionActivate?.license;
export const getErrorsAsData = ({ data } = {}) => data?.gitlabSubscriptionActivate?.errors || [];

export const updateSubscriptionAppCache = (cache, mutation) => {
  const license = getLicenseFromData(mutation);
  if (!license) {
    return;
  }
  const { query } = subscriptionQueries;
  const { query: historyQuery } = subscriptionHistoryQueries;
  const data = produce({}, (draftData) => {
    draftData.currentLicense = license;
  });
  cache.writeQuery({ query, data });
  const subscriptionsList = cache.readQuery({ query: historyQuery });
  const subscriptionListData = produce(subscriptionsList, (draftData) => {
    draftData.licenseHistoryEntries.nodes = [
      license,
      ...subscriptionsList.licenseHistoryEntries.nodes,
    ];
  });
  cache.writeQuery({ query: historyQuery, data: subscriptionListData });
};
