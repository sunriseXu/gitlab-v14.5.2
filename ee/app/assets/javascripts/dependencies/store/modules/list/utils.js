import { REPORT_STATUS } from './constants';

export const hasDependencyList = ({ dependencies }) => Array.isArray(dependencies);

export const hasReportStatus = ({ report }) =>
  Boolean(report && Object.values(REPORT_STATUS).includes(report.status));

export const isValidResponse = ({ data }) =>
  Boolean(data && hasDependencyList(data) && hasReportStatus(data));
