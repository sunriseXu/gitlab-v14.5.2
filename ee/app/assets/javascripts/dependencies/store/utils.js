import listModule from './modules/list';

export const addListType = (store, listType) => {
  const { initialState, namespace } = listType;
  store.registerModule(namespace, listModule());
  store.dispatch('addListType', listType);
  store.dispatch(`${namespace}/setInitialState`, initialState);
};
