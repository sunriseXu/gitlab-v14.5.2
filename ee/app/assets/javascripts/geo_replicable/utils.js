import createGqClient, { fetchPolicies } from '~/lib/graphql';

/*
  This file uses a NO_CACHE policy due to the need for Geo data to always be fresh.
  The UI this serves is used to watch the "syncing" process of items and their statuses
  will need to be constantly re-queried as the user navigates around to not mistakenly
  think the sync process is broken.
*/

export const gqClient = createGqClient(
  {},
  {
    fetchPolicy: fetchPolicies.NO_CACHE,
  },
);
