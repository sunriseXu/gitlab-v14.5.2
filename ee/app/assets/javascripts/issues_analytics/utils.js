/**
 * This util method takes the global page filters and transforms parameters which
 * are not standardized between the internal issue analytics api and the public
 * issues api.
 *
 * @param {Object} filters the global filters used to fetch issues data
 *
 * @returns {Object} the transformed filters for the public api
 */
export const transformFilters = ({
  label_name: labels = null,
  milestone_title: milestone = null,
  ...restOfFilters
}) => ({
  ...restOfFilters,
  labels,
  milestone,
});
