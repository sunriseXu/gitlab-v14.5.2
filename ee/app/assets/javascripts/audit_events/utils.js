import { parsePikadayDate, pikadayToString } from '~/lib/utils/datetime_utility';
import { AVAILABLE_TOKEN_TYPES, AUDIT_FILTER_CONFIGS, ENTITY_TYPES } from './constants';
import { parseUsername, displayUsername } from './token_utils';

export const getTypeFromEntityType = (entityType) => {
  return AUDIT_FILTER_CONFIGS.find(
    ({ entityType: configEntityType }) => configEntityType === entityType,
  )?.type;
};

export const getEntityTypeFromType = (type) => {
  return AUDIT_FILTER_CONFIGS.find(({ type: configType }) => configType === type)?.entityType;
};

export const parseAuditEventSearchQuery = ({
  created_after: createdAfter,
  created_before: createdBefore,
  entity_type: entityType,
  entity_username: entityUsername,
  author_username: authorUsername,
  ...restOfParams
}) => ({
  ...restOfParams,
  created_after: createdAfter ? parsePikadayDate(createdAfter) : null,
  created_before: createdBefore ? parsePikadayDate(createdBefore) : null,
  entity_type: getTypeFromEntityType(entityType),
  entity_username: displayUsername(entityUsername),
  author_username: displayUsername(authorUsername),
});

export const createAuditEventSearchQuery = ({ filterValue, startDate, endDate, sortBy }) => {
  const entityValue = filterValue.find((value) => AVAILABLE_TOKEN_TYPES.includes(value.type));
  const entityType = getEntityTypeFromType(entityValue?.type);
  const filterData = entityValue?.value.data;

  const params = {
    created_after: startDate ? pikadayToString(startDate) : null,
    created_before: endDate ? pikadayToString(endDate) : null,
    sort: sortBy,
    entity_type: entityType,
    entity_id: null,
    entity_username: null,
    author_username: null,
    // When changing the search parameters, we should be resetting to the first page
    page: null,
  };

  switch (entityType) {
    case ENTITY_TYPES.USER:
      params.entity_username = parseUsername(filterData);
      break;
    case ENTITY_TYPES.AUTHOR:
      params.author_username = parseUsername(filterData);
      break;
    default:
      params.entity_id = filterData;
  }

  return params;
};
