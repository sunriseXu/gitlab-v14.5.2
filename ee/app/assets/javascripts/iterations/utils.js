import { formatDate } from '~/lib/utils/datetime_utility';

const PERIOD_DATE_FORMAT = 'mmm d, yyyy';

/**
 * The arguments are two date strings in formatted in ISO 8601 (YYYY-MM-DD)
 *
 * @returns {string} ex. "Oct 1, 2021 - Oct 10, 2021"
 */
export function getIterationPeriod({ startDate, dueDate }) {
  const start = formatDate(startDate, PERIOD_DATE_FORMAT, true);
  const due = formatDate(dueDate, PERIOD_DATE_FORMAT, true);
  return `${start} - ${due}`;
}
