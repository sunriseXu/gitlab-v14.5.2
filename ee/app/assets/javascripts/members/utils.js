import { __ } from '~/locale';
import { generateBadges as CEGenerateBadges, isDirectMember } from '~/members/utils';

export {
  isGroup,
  isDirectMember,
  isCurrentUser,
  canRemove,
  canResend,
  canUpdate,
} from '~/members/utils';

export const generateBadges = ({ member, isCurrentUser, canManageMembers }) => [
  ...CEGenerateBadges({ member, isCurrentUser, canManageMembers }),
  {
    show: member.usingLicense,
    text: __('Is using seat'),
    variant: 'neutral',
  },
  {
    show: member.groupSso,
    text: __('SAML'),
    variant: 'info',
  },
  {
    show: member.groupManagedAccount,
    text: __('Managed Account'),
    variant: 'info',
  },
  {
    show: member.canOverride,
    text: __('LDAP'),
    variant: 'info',
  },
  {
    show: member.provisionedByThisGroup,
    text: __('Enterprise'),
    variant: 'info',
  },
];

export const canOverride = (member) => member.canOverride && isDirectMember(member);
