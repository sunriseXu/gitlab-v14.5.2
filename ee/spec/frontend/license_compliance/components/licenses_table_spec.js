import { shallowMount } from '@vue/test-utils';
import LicensesTable from 'ee/license_compliance/components/licenses_table.vue';
import LicensesTableRow from 'ee/license_compliance/components/licenses_table_row.vue';
import { makeLicense } from './utils';

describe('LicensesTable component', () => {
  let wrapper;

  const factory = (propsData = {}) => {
    wrapper = shallowMount(LicensesTable, {
      propsData: { ...propsData },
    });
  };

  const findTableRowHeader = () => wrapper.find('.table-row-header');
  const findRows = () => wrapper.findAll(LicensesTableRow);

  afterEach(() => {
    wrapper.destroy();
  });

  describe('given an empty list of licenses', () => {
    beforeEach(() => {
      factory({
        licenses: [],
        isLoading: false,
      });
    });

    it('renders the table headers', () => {
      expect(findTableRowHeader().element).toMatchSnapshot();
    });

    it('renders the empty license table', () => {
      expect(findRows()).toHaveLength(0);
    });
  });

  [true, false].forEach((isLoading) => {
    describe(`given a list of licenses (${isLoading ? 'loading' : 'loaded'})`, () => {
      let licenses;
      beforeEach(() => {
        licenses = [makeLicense(), makeLicense({ name: 'foo' })];

        factory({
          licenses,
          isLoading,
        });
      });

      it('renders the table headers', () => {
        expect(findTableRowHeader().element).toMatchSnapshot();
      });

      it('passes the correct props to the table rows', () => {
        expect(findRows()).toHaveLength(licenses.length);
        expect(findRows().wrappers.map((x) => x.props())).toEqual(
          licenses.map((license) => ({
            license,
            isLoading,
          })),
        );
      });
    });
  });
});
