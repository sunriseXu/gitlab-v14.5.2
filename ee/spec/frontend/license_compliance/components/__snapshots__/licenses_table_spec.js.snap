// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LicensesTable component given a list of licenses (loaded) renders the table headers 1`] = `
<div
  class="gl-responsive-table-row table-row-header text-2 bg-secondary-50 px-2"
  role="row"
>
  <div
    class="table-section section-30"
    role="rowheader"
  >
    
      Name
    
  </div>
  <div
    class="table-section section-70"
    role="rowheader"
  >
    
      Component
    
  </div>
</div>
`;

exports[`LicensesTable component given a list of licenses (loading) renders the table headers 1`] = `
<div
  class="gl-responsive-table-row table-row-header text-2 bg-secondary-50 px-2"
  role="row"
>
  <div
    class="table-section section-30"
    role="rowheader"
  >
    
      Name
    
  </div>
  <div
    class="table-section section-70"
    role="rowheader"
  >
    
      Component
    
  </div>
</div>
`;

exports[`LicensesTable component given an empty list of licenses renders the table headers 1`] = `
<div
  class="gl-responsive-table-row table-row-header text-2 bg-secondary-50 px-2"
  role="row"
>
  <div
    class="table-section section-30"
    role="rowheader"
  >
    
      Name
    
  </div>
  <div
    class="table-section section-70"
    role="rowheader"
  >
    
      Component
    
  </div>
</div>
`;
